File: python/02_increment_shared_counter.py
import threading

counter = 0
lock = threading.Lock()

def increment():
    global counter
    with lock:
        counter += 1
    print(f"Counter value: {counter}")

threads = []
for i in range(10):
    thread = threading.Thread(target=increment)
    thread.start()
    threads.append(thread)

for thread in threads:
    thread.join()

print("All threads completed")
print(f"Final counter value: {counter}")

File: C#/02_increment_shared_counter.cs
using System;
using System.Threading;

class Program
{
    static int counter = 0;
    static readonly object lockObject = new object();

    static void Increment()
    {
        lock (lockObject)
        {
            counter++;
        }
        Console.WriteLine("Counter value: {0}", counter);
    }

    static void Main(string[] args)
    {
        Thread[] threads = new Thread[10];
        for (int i = 0; i < threads.Length; i++)
        {
            threads[i] = new Thread(Increment);
            threads[i].Start();
        }

        foreach (Thread thread in threads)
        {
            thread.Join();
        }

        Console.WriteLine("All threads completed");
        Console.WriteLine("Final counter value: {0}", counter);
    }
}
